proposals/0021-vk-coop-matrix.md

```diff
@@ -0,0 +1,457 @@
+  CooperativeMatrix operator*(ComponentType scalar);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to
+  // data using the given memory layout, stride, and memory access mask.
+  // `NonPrivatePointer` and `MakePointerVisible` with the workgroup scope
+  // will be added to the memory access mask to make the memory coherent.
+  //
+  // This function uses a SPIR-V pointer because HLSL does not allow grouphsared
+  // memory object to be passed by reference. The pointer is a hack to get
+  // around that.
+  template <MemoryAccessMask memoryAccessMask, CooperativeMatrixLayout layout,
+            class Type>
+  void Store(WorkgroupSpirvPointer<Type> data, uint32_t stride);
+
+  // Same as above, but uses MemoryAccessMaskNone for the memory access mask.
+  template <CooperativeMatrixLayout layout, class Type>
+  void Store(WorkgroupSpirvPointer<Type> data, uint32_t stride);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to
+  // data[index] using the given memory layout, stride, and memory access mask.
+  // `NonPrivatePointer` and `MakePointerVisible` with the QueueFamily scope
+  // will be added to the memory access mask to make the memory coherent.
+  //
+  template <MemoryAccessMask memoryAccessMask, CooperativeMatrixLayout layout,
+            class Type>
+  void Store(globallycoherent RWStructuredBuffer<Type> data, uint32_t index,
+             uint32_t stride);
+
+  // Same as above, but uses MemoryAccessMaskNone for the memory access mask.
+  template <CooperativeMatrixLayout layout, class Type>
+  void Store(globallycoherent RWStructuredBuffer<Type> data, uint32_t index,
+             uint32_t stride);
+
+  // Loads a cooperative matrix using OpCooperativeMatrixStoreKHR from
+  // data using the given memory layout, stride, and memory access mask.
+  // `NonPrivatePointer` and `MakePointerAvailable` with the workgroup scope
+  // will be added to the memory access mask to make the memory coherent.
+  //
+  // This function uses a SPIR-V pointer because HLSL does not allow grouphsared
+  // memory object to be passed by reference. The pointer is a hack to get
+  // around that.
+  template <MemoryAccessMask memoryAccessMask, CooperativeMatrixLayout layout,
+            class Type>
+  static CooperativeMatrix Load(WorkgroupSpirvPointer<Type> data,
+                                uint32_t stride);
+
+  // Same as above, but uses MemoryAccessMaskNone for the memory access mask.
+  template <CooperativeMatrixLayout layout, class Type>
+  static CooperativeMatrix Load(WorkgroupSpirvPointer<Type> data,
+                                uint32_t stride);
+
+  // Loads a cooperative matrix using OpCooperativeMatrixLoadKHR from
+  // data[index] using the given memory layout, stride, and memory access mask.
+  // `NonPrivatePointer` and `MakePointerAvailable` with the QueueFamily scope
+  // will be added to the memory access mask to make the memory coherent.
```

Comment:
Is this saying they're unconditionally coherent? Most of the time you want noncoherent, for performance. In the GLSL extension, you get coherent if and only if the buffer/member is qualified with the coherent keyword.

Reply:
I cannot have an overload distinguishes between something marked globallycoherent or not. If we wanted both, I'd have to have two separate functions with different names.

I went with this because all of the `spv.coopmat.*` tests in glslang us coherent for everything, I assumed that was the common case.

Let me know the best implementation. You know better how it will be used. I'm just implementing it. 

FYI: I interpreted that the workgroup memory should always be coherent from the GLSL tests:

https://github.com/KhronosGroup/glslang/blob/7c4d91e7819a1d27213aa3499953d54ae1a00e8f/Test/spv.coopmat.comp#L112-L113

The expected result is

https://github.com/KhronosGroup/glslang/blob/7c4d91e7819a1d27213aa3499953d54ae1a00e8f/Test/baseResults/spv.coopmat.comp.out#L352C1-L353C112

Reply:
> I went with this because all of the spv.coopmat.* tests in glslang us coherent for everything, I assumed that was the common case.

That was probably just me testing that the right qualifier was emitted and then copy/pasting it everywhere.

Can you just let the shader author choose by setting the right bits in memoryAccessMask? Maybe automatically add it for shared memory because it's free-ish.

> FYI: I interpreted that the workgroup memory should always be coherent from the GLSL tests:

Right, the GLSL spec says "Shared variables are implicitly coherent".

Reply:
For RWStructuredBuffers, I added a CoherentLoad and CoherentStore. The problem with just letting people set the bits is that it gets ugly because of the scope argument or other extra arguments depending on which memory operands are used.

Reply:
Right, that makes sense. Adding an overload is fine with me.